### PR TITLE
Add VlanId in the cmdAdd Result struct 

### PR DESCRIPTION
**Pull request by cgchinmay**
This VlanId will appear in the prevResult during cmdDel request
The cni can then fetch the vlanId and delete the network associated with vlanId
This will help prevent calls made to api server from ipamd to fetch the vlan device information 

#1644 
#1313 

Tested cni and ipamd integration tests
Test with e2e tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jayanthvn** [comment]
Can you please fix the conflict - test/framework/resources/k8s/utils/daemonset.go

**Comment by jayanthvn** [comment]
@cgchinmay - Even here branch needs update.

File: cmd/routed-eni-cni-plugin/cni.go [hunk 1]
```diff
@@ -192,16 +204,20 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
...
+	var dummyVlanInterface *current.Interface
 	if r.PodVlanId != 0 {
-		hostVethName = generateHostVethName("vlan", string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
-
+		hostVethName = generateHostVethName(vlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
 		err = driverClient.SetupPodENINetwork(hostVethName, args.IfName, args.Netns, v4Addr, v6Addr, int(r.PodVlanId), r.PodENIMAC,
 			r.PodENISubnetGW, int(r.ParentIfIndex), mtu, log)
+		// This is a dummyVlanInterface is created only to send podVlanId information as a part of the Result struct
+		// The podVlanId is used by DEL cmd fetched from the prevResult struct to cleanup pod network
+		dummyVlanInterfaceName := generateHostVethName(dummyVlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
```
Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:214]: As @achevuru mentioned, let's document on why we added the dummyVlanInterfaceName for future reference.

File: cmd/routed-eni-cni-plugin/cni.go [hunk 2]
```diff
@@ -122,6 +132,8 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
 	log.Infof("Received CNI add request: ContainerID(%s) Netns(%s) IfName(%s) Args(%s) Path(%s) argsStdinData(%s)",
 		args.ContainerID, args.Netns, args.IfName, args.Args, args.Path, args.StdinData)
 
+	log.Infof("Prev Result: %v\n", conf.PrevResult)
```
Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:135]: nit: Can we make this Debugf?

File: cmd/routed-eni-cni-plugin/cni.go [hunk 3]
```diff
@@ -192,16 +204,20 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
...
-		hostVethName = generateHostVethName("vlan", string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
-
+		hostVethName = generateHostVethName(vlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
 		err = driverClient.SetupPodENINetwork(hostVethName, args.IfName, args.Netns, v4Addr, v6Addr, int(r.PodVlanId), r.PodENIMAC,
 			r.PodENISubnetGW, int(r.ParentIfIndex), mtu, log)
+		// This is a dummyVlanInterface is created only to send podVlanId information as a part of the Result struct
+		// The podVlanId is used by DEL cmd fetched from the prevResult struct to cleanup pod network
+		dummyVlanInterfaceName := generateHostVethName(dummyVlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
+		dummyVlanInterface = &current.Interface{Name: dummyVlanInterfaceName, Mac: fmt.Sprint(r.PodVlanId)}
+		log.Infof("Using dummy vlanInterface: %v", dummyVlanInterface)
```
Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:216]: Even this, lets make it Debugf

File: cmd/routed-eni-cni-plugin/cni.go [hunk 4]
```diff
@@ -283,6 +306,40 @@ func del(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
...
+		dummyVlanInterfaceName := generateHostVethName(dummyVlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
+		for _, iface := range prevResult.Interfaces {
+			if iface.Name == dummyVlanInterfaceName {
+				podVlanId, err := strconv.Atoi(iface.Mac)
+				if err != nil {
+					return errors.Wrap(err, "Failed to parse vlanId from prevResult")
+				}
+				// podVlanId == 0 means pod is not using branch ENI
+				// then fallback to existing cleanup
+				if podVlanId == 0 {
```
Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:328]: Will we have this situation? I mean dummyVlanInterfaceName is added if SGPP is enabled so we should typically have podVlanID right? if it is 0 then it is an error.

File: cmd/routed-eni-cni-plugin/cni.go [hunk 5]
```diff
@@ -283,6 +306,40 @@ func del(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
 		return errors.Wrap(err, "del cmd: failed to load k8s config from args")
 	}
 
+	if args.Netns == "" {
```
Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:309]: Is there a specific reason this check is added?

File: misc/10-aws.conflist [hunk 6]
```diff
@@ -1,6 +1,7 @@
 {
-  "cniVersion": "0.3.1",
+  "cniVersion": "0.4.0",
   "name": "aws-cni",
+  "disableCheck": true,
```
Comment (jayanthvn) [inline misc/10-aws.conflist:4]: Can you enable bandwidth plugin and just verify if the conflist file is regenerated properly?

File: cmd/routed-eni-cni-plugin/cni.go (hunk 5 above)
Comment (cgchinmay) [reply cmd/routed-eni-cni-plugin/cni.go:321]: Added comments

File: cmd/routed-eni-cni-plugin/cni.go (hunk 4 above)
Comment (cgchinmay) [reply cmd/routed-eni-cni-plugin/cni.go:342]: Updated

File: cmd/routed-eni-cni-plugin/cni.go (hunk 1 above)
Comment (cgchinmay) [reply cmd/routed-eni-cni-plugin/cni.go:219]: Updated comments

File: misc/10-aws.conflist (hunk 6 above)
Comment (cgchinmay) [reply misc/10-aws.conflist:4]: Verified 
```
{
  "cniVersion": "0.4.0",
  "name": "aws-cni",
  "disableCheck": true,
  "plugins": [
    {
      "name": "aws-cni",
      "type": "aws-cni",
      "vethPrefix": "eni",
      "mtu": "9001",
      "pluginLogFile": "/var/log/aws-routed-eni/plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "name": "egress-v4-cni",
      "type": "egress-v4-cni",
      "mtu": 9001,
      "enabled": "false",
      "nodeIP": "192.168.87.116",
      "ipam": {
        "type": "host-local",
        "ranges": [
          [
            {
              "subnet": "169.254.172.0/22"
            }
          ]
        ],
        "routes": [
          {
            "dst": "0.0.0.0/0"
          }
        ],
        "dataDir": "/run/cni/v6pd/egress-v4-ipam"
      },
      "pluginLogFile": "/var/log/aws-routed-eni/egress-v4-plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "type": "portmap",
      "capabilities": {
        "portMappings": true
      },
      "snat": true
    },
    {
      "type": "bandwidth",
      "capabilities": {
        "bandwidth": true
      }
    }
  ]
}
```

Comment (jayanthvn) [reply misc/10-aws.conflist:4]: Thanks for verifying

**Review by jayanthvn — APPROVED** [review]
lgtm

File: cmd/routed-eni-cni-plugin/cni.go (hunk 4 above)
Comment (jayanthvn) [reply cmd/routed-eni-cni-plugin/cni.go:342]: If we have an unsupported instance and PPSG is enabled then we can actually have podvlanID as 0. Maybe we need to special case this? @achevuru any thoughts?

Comment (achevuru) [reply cmd/routed-eni-cni-plugin/cni.go:342]: I see that we only add the `dummyVlanInterface` if `podVlanId != 0` during the `add` flow. So, I think we might be Ok here in the del flow as this is executed only if we find `dummyVlanInterface` in prevResult...

**Pull request merged**